File: rust/src/parser.rs
//! zser message parser

#[cfg(not(feature = "std"))]
use collections::Vec;
use errors::*;
use varint;

/// Default maximum length of a zser message: 1kB
/// This is conservative as zser's main intended use case is a credential format
pub const DEFAULT_MAX_LENGTH: usize = 1024;

/// Default maximum depth (i.e. default max level of nested messages)
pub const DEFAULT_MAX_DEPTH: usize = 8;

/// Parser for zser messages
pub struct Parser<'m, H: Handler> {
    /// Maximum length message we'll accept
    max_length: usize,

    /// Maximum depth of nested messages allowed
    max_depth: usize,

    /// Bodies of nested messages remaining to be processed
    remaining: Vec<&'m [u8]>,

    /// Callbacks to invoke to construct the resulting type
    handler: H,
}

impl<'m, H: Handler> Parser<'m, H> {
    /// Create a new Parser
    pub fn new(handler: H) -> Parser<'m, H> {
        Parser {
            max_length: DEFAULT_MAX_LENGTH,
            max_depth: DEFAULT_MAX_DEPTH,
            remaining: Vec::new(),
            handler: handler,
        }
    }

    /// Parse the given zser message, invoking callbacks as necessary
    pub fn parse(&mut self, message: &'m [u8]) -> Result<()> {
        if message.len() > self.max_length {
            return Err(ErrorKind::OversizedMessage(message.len(), self.max_length).into());
        }

        if self.remaining.len() >= self.max_depth {
            return Err(ErrorKind::MaxDepthExceeded(self.max_depth).into());
        }

        self.remaining.push(message);

        while !self.remaining.last().expect("remaining").is_empty() {
            let (id, wiretype) = self.parse_field_prefix()?;

            match wiretype {
                0 => self.parse_u64(id)?,
                2 => self.parse_message(id)?,
                3 => self.parse_binary(id)?,
                _ => return Err(ErrorKind::UnknownWiretype(wiretype).into()),
            }
        }

        // In theory, this should never panic
        self.remaining.pop().expect("empty remaining stack");

        Ok(())
    }

    /// Finish parsing, returning the resulting object produced by the builder
    pub fn finish(self) -> Result<H::T> {
        if self.remaining.is_empty() {
            Ok(self.handler.finish())
        } else {
            Err(ErrorKind::UnconsumedMessages(self.remaining.len()).into())
        }
    }

    /// Pop the top item in the remaining stack and parse a varint from it
    pub fn parse_varint(&mut self) -> Result<(u64, &'m [u8])> {
        let slice = self.remaining.pop().expect("remaining slice");
        let mut slice_ref = &slice[..];
        let value = varint::decode(&mut slice_ref)?;

        Ok((value, slice_ref))
    }

    /// Parse the integer each field starts with, extracting field ID and wiretype
    pub fn parse_field_prefix(&mut self) -> Result<(u64, u64)> {
        let (value, remaining) = self.parse_varint()?;
        self.remaining.push(remaining);

        let field_id = value >> 3;
        let wiretype = value & 0x7;

        Ok((field_id, wiretype))
    }

    /// Parse a u64 value stored as a prefix varint
    pub fn parse_u64(&mut self, id: u64) -> Result<()> {
        let (value, remaining) = self.parse_varint()?;
        self.remaining.push(remaining);
        self.handler.uint64(id, value);

        Ok(())
    }

    /// Parse a blob of data that begins with a length prefix
    pub fn parse_length_prefixed_data(&mut self) -> Result<&'m [u8]> {
        let (length_u64, remaining) = self.parse_varint()?;
        let length = length_u64 as usize;

        if remaining.len() < length {
            let message = format!("want {} bytes, have {}", length, remaining.len());
            return Err(ErrorKind::TruncatedMessage(message).into());
        }

        let result = &remaining[..length];
        self.remaining.push(&remaining[length..]);

        Ok(result)
    }

    /// Parse a nested message
    pub fn parse_message(&mut self, id: u64) -> Result<()> {
        self.handler.begin_nested();

        let nested_message = self.parse_length_prefixed_data()?;
        self.parse(nested_message)?;
        self.handler.end_nested(id);

        Ok(())
    }

    /// Parse a field containing binary data
    pub fn parse_binary(&mut self, id: u64) -> Result<()> {
        let data = self.parse_length_prefixed_data()?;
        self.handler.binary(id, data);

        Ok(())
    }
}

/// Callback API used by the parser to process parsed data
// TODO: switch to serde
pub trait Handler {
    /// Type produced by this handler when parsing is complete
    type T;

    /// Called when a uint64 value with the given field ID is parsed
    fn uint64(&mut self, id: u64, value: u64);

    /// Called when we've received binary data with the given ID
    fn binary(&mut self, id: u64, data: &[u8]);

    /// Indicate we've entered a new nested message
    fn begin_nested(&mut self);

    /// Indicate we've reached the end of a nested message with the given ID
    fn end_nested(&mut self, id: u64);

    /// Return the fully parsed object
    fn finish(self) -> Self::T;
}

File: go/parser.go
// parser.go: veriform message parser

package veriform

import (
	"bytes"
	"fmt"
)

// Default maximum length of a veriform message: 1kB
// This is conservative as veriform's main intended use case is a credential format
const DEFAULT_MAX_LENGTH = 1024

// Default maximum depth (i.e. default max level of nested messages)
const DEFAULT_MAX_DEPTH = 8

// Parser for veriform messages
type parser struct {
	// Maximum length message we'll accept
	maxLength uint

	// Maximum depth of nested messages allowed
	maxDepth uint

	// Bodies of nested messages remaining to be processed
	remaining [][]byte

	// Callbacks to invoke to construct the resulting type
	callbacks handler
}

// Create a new Parser
func NewParser(callbacks handler) *parser {
	return &parser{
		DEFAULT_MAX_LENGTH,
		DEFAULT_MAX_DEPTH,
		make([][]byte, 0),
		callbacks,
	}
}

// Parse the given veriform message, invoking callbacks as necessary
func (p *parser) Parse(message []byte) error {
	if len(message) > int(p.maxLength) {
		return fmt.Errorf("oversized message: %d bytes (max %d)", len(message), p.maxLength)
	}

	if len(p.remaining) >= int(p.maxDepth) {
		return fmt.Errorf("max depth of %d nested messages exceeded", p.maxDepth)
	}

	p.remaining = append(p.remaining, message)

	for len(p.remaining[len(p.remaining)-1]) > 0 {
		fieldID, wireType, err := p.parseFieldPrefix()
		if err != nil {
			return err
		}

		switch wireType {
		case 0:
			err = p.parseUint64(fieldID)
		case 2:
			err = p.parseMessage(fieldID)
		case 3:
			err = p.parseBytes(fieldID)
		default:
			err = fmt.Errorf("unknown wiretype: %d", wireType)
		}

		if err != nil {
			return err
		}
	}

	p.remaining = p.remaining[:len(p.remaining)-1]

	return nil
}

// Finish parsing, returning the resulting object produced by the builder
func (p *parser) Finish() (interface{}, error) {
	if len(p.remaining) == 0 {
		return p.callbacks.Finish(), nil
	} else {
		return nil, fmt.Errorf("not finished parsing: %d messages in buffer", len(p.remaining))
	}
}

// Pop the top item in the remaining stack and parse a varint from it
// TODO: better integrate io.Reader to avoid unnecessary slicing
func (p *parser) parseVarint() (uint64, []byte, error) {
	slice := p.remaining[len(p.remaining)-1]
	reader := bytes.NewReader(slice)
	p.remaining = p.remaining[:len(p.remaining)-1]

	value, err := DecodeVarint(reader)
	if err != nil {
		return 0, nil, err
	}

	return value, slice[len(slice)-reader.Len():], nil
}

// Parse the integer each field starts with, extracting field ID and wiretype
func (p *parser) parseFieldPrefix() (FieldID, WireType, error) {
	value, remaining, err := p.parseVarint()
	if err != nil {
		return 0, 0, err
	}

	p.remaining = append(p.remaining, remaining)

	fieldID := FieldID(value >> 3)
	wireType := WireType(value & 0x7)

	return fieldID, wireType, nil
}

// Parse a u64 value stored as a prefix varint
func (p *parser) parseUint64(fieldID FieldID) error {
	value, remaining, err := p.parseVarint()
	if err != nil {
		return err
	}

	p.remaining = append(p.remaining, remaining)
	p.callbacks.Uint64(fieldID, value)

	return nil
}

// Parse a blob of data that begins with a length prefix
func (p *parser) parseLengthPrefixedData() ([]byte, error) {
	length, remaining, err := p.parseVarint()
	if err != nil {
		return nil, err
	}

	if len(remaining) < int(length) {
		return nil, fmt.Errorf("message truncated: want %d bytes (have %d)", length, len(remaining))
	}

	result := remaining[:length]
	p.remaining = append(p.remaining, remaining[length:])

	return result, nil
}

// Parse a nested message
func (p *parser) parseMessage(fieldID FieldID) error {
	p.callbacks.BeginNested()

	nestedMessage, err := p.parseLengthPrefixedData()
	if err != nil {
		return err
	}

	err = p.Parse(nestedMessage)
	if err != nil {
		return err
	}

	p.callbacks.EndNested(fieldID)
	return nil
}

// Parse a field containing binary data
func (p *parser) parseBytes(fieldID FieldID) error {
	data, err := p.parseLengthPrefixedData()
	if err != nil {
		return err
	}

	p.callbacks.Bytes(fieldID, data)
	return nil
}

// Callback API used by the parser to process parsed data
type handler interface {
	// Called when a uint64 value with the given field ID is parsed
	Uint64(fieldID FieldID, value uint64)

	// Called when we've received binary data with the given ID
	Bytes(fieldID FieldID, data []byte)

	// Indicate we've entered a new nested message
	BeginNested()

	// Indicate we've reached the end of a nested message with the given ID
	EndNested(fieldID FieldID)

	// Return the fully parsed object
	Finish() interface{}
}
